atlasdb-client/src/main/java/com/palantir/atlasdb/cache/OffHeapTimestampCache.java

```diff
@@ -0,0 +1,156 @@
+import com.palantir.logsafe.exceptions.SafeIllegalArgumentException;
+
+public final class OffHeapTimestampCache implements TimestampCache {
+    private static final String TIMESTAMP_CACHE_NAMESPACE = "timestamp_cache";
+    private static final String BATCHER_PURPOSE = "off-heap-timestamp-cache";
+    private static final Logger log = LoggerFactory.getLogger(OffHeapTimestampCache.class);
+
+    private final PersistentTimestampStore persistentTimestampStore;
+    private final int maxSize;
+    private final AtomicReference<CacheDescriptor> cacheDescriptor = new AtomicReference<>();
+    private final ConcurrentMap<Long, Long> concurrentHashMap = new ConcurrentHashMap<>();
+    private final DisruptorAutobatcher<Map.Entry<Long, Long>, Map.Entry<Long, Long>> autobatcher;
+
+    public static TimestampCache create(PersistentTimestampStore persistentTimestampStore, int maxSize) {
+        StoreNamespace storeNamespace = persistentTimestampStore.createNamespace(TIMESTAMP_CACHE_NAMESPACE);
+
+        CacheDescriptor cacheDescriptor = ImmutableCacheDescriptor.builder()
+                .currentSize(new AtomicInteger())
+                .storeNamespace(storeNamespace)
+                .build();
+
+        return new OffHeapTimestampCache(persistentTimestampStore, cacheDescriptor, maxSize);
+    }
+
+    private OffHeapTimestampCache(
+            PersistentTimestampStore persistentTimestampStore,
+            CacheDescriptor cacheDescriptor,
+            int maxSize) {
+        this.persistentTimestampStore = persistentTimestampStore;
+        this.cacheDescriptor.set(cacheDescriptor);
+        this.maxSize = maxSize;
+        this.autobatcher = Autobatchers.coalescing(new WriteBatcher(this))
+                .safeLoggablePurpose(BATCHER_PURPOSE)
+                .build();
+    }
+
+    @Override
+    public void clear() {
+        CacheDescriptor proposedCacheDescriptor = constructCacheProposal(persistentTimestampStore);
+
+        CacheDescriptor previous = cacheDescriptor.getAndUpdate(prev -> proposedCacheDescriptor);
+        if (previous != null) {
+            persistentTimestampStore.dropNamespace(previous.storeNamespace());
+        }
+    }
+
+
+    @Override
+    public void putAlreadyCommittedTransaction(Long startTimestamp, Long commitTimestamp) {
+        if (concurrentHashMap.putIfAbsent(startTimestamp, commitTimestamp) != null) {
+            return;
+        }
+        Futures.getUnchecked(autobatcher.apply(Maps.immutableEntry(startTimestamp, commitTimestamp)));
+    }
+
+    @Nullable
+    @Override
+    public Long getCommitTimestampIfPresent(Long startTimestamp) {
+        Long value = concurrentHashMap.get(startTimestamp);
+        if (value != null) {
+            return value;
+        }
+
+        return persistentTimestampStore.get(cacheDescriptor.get().storeNamespace(), startTimestamp);
+    }
+
+    private static CacheDescriptor constructCacheProposal(PersistentTimestampStore persistentTimestampStore) {
```

Comment:
nit: would suggest renaming, maybe to `createNamespaceAndConstructCacheProposal` since this has side effects

Reply:
Renamed.